### PR TITLE
New Resource: `azurerm_monitor_data_collection_rule_association`

### PR DESCRIPTION
**Pull request by teowa**
### New resource `azurerm_monitor_data_collection_rule_association`
Resolves https://github.com/hashicorp/terraform-provider-azurerm/issues/17831
Reference:
1. [Microsoft doc](https://docs.microsoft.com/en-us/azure/azure-monitor/essentials/data-collection-rule-overview)
2. [REST API specification 2021-04-01](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/monitor/resource-manager/Microsoft.Insights/stable/2021-04-01/dataCollectionRuleAssociations_API.json)

<details>
<summary>Test Result</summary>

```
make acctests SERVICE='monitor' TESTARGS='-run=TestAccMonitorDataCollectionRuleAssociation_' TESTTIMEOUT='60m'
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./internal/services/monitor -run=TestAccMonitorDataCollectionRuleAssociation_ -timeout 60m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccMonitorDataCollectionRuleAssociation_basic
=== PAUSE TestAccMonitorDataCollectionRuleAssociation_basic
=== RUN   TestAccMonitorDataCollectionRuleAssociation_basicEndpoint
=== PAUSE TestAccMonitorDataCollectionRuleAssociation_basicEndpoint
=== RUN   TestAccMonitorDataCollectionRuleAssociation_requiresImport
=== PAUSE TestAccMonitorDataCollectionRuleAssociation_requiresImport
=== RUN   TestAccMonitorDataCollectionRuleAssociation_update
=== PAUSE TestAccMonitorDataCollectionRuleAssociation_update
=== RUN   TestAccMonitorDataCollectionRuleAssociation_updateDataCollectionRuleId
=== PAUSE TestAccMonitorDataCollectionRuleAssociation_updateDataCollectionRuleId
=== RUN   TestAccMonitorDataCollectionRuleAssociation_complete
=== PAUSE TestAccMonitorDataCollectionRuleAssociation_complete
=== CONT  TestAccMonitorDataCollectionRuleAssociation_basic
=== CONT  TestAccMonitorDataCollectionRuleAssociation_update
=== CONT  TestAccMonitorDataCollectionRuleAssociation_requiresImport
=== CONT  TestAccMonitorDataCollectionRuleAssociation_basicEndpoint
=== CONT  TestAccMonitorDataCollectionRuleAssociation_complete
=== CONT  TestAccMonitorDataCollectionRuleAssociation_updateDataCollectionRuleId
--- PASS: TestAccMonitorDataCollectionRuleAssociation_complete (342.13s)
--- PASS: TestAccMonitorDataCollectionRuleAssociation_basic (343.31s)
--- PASS: TestAccMonitorDataCollectionRuleAssociation_basicEndpoint (350.74s)
--- PASS: TestAccMonitorDataCollectionRuleAssociation_requiresImport (364.89s)
--- PASS: TestAccMonitorDataCollectionRuleAssociation_updateDataCollectionRuleId (431.38s)
--- PASS: TestAccMonitorDataCollectionRuleAssociation_update (519.71s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/monitor       519.745s

```

</details>

### CODE REVIEW

File: website/docs/r/monitor_data_collection_rule_association.html.markdown [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+---
+
+* `data_collection_endpoint_id` - (Optional) The ID of the Data Collection Endpoint which will be associated to the target resource.
+
+* `data_collection_rule_id` - (Optional) The ID of the Data Collection Rule which will be associated to the target resource.
+ 
+-> **NOTE** Exactly one of `data_collection_endpoint_id` and `data_collection_rule_id` blocks must be specified.
+
+-> **NOTE** An association of Data Collection Endpoint must be named 'configurationAccessEndpoint'.
```
Comment (katbyte) [inline website/docs/r/monitor_data_collection_rule_association.html.markdown:129]: this should be below the name property?

File: website/docs/r/monitor_data_collection_rule_association.html.markdown [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+  description                 = "example"
+}
+
+```
+
+## Arguments Reference
+
+The following arguments are supported:
+
+* `name` - (Required) The name which should be used for this Data Collection Rule. Changing this forces a new Data Collection Rule to be created.
```
Comment (katbyte) [inline website/docs/r/monitor_data_collection_rule_association.html.markdown:117]: if this cannot be configured to anything but configurationAccessEndpoint should we even be exposing the name? maybe we should remove it>

**Review by katbyte — CHANGES_REQUESTED** [review]

File: website/docs/r/monitor_data_collection_rule_association.html.markdown (hunk 1 above)
Comment (teowa) [reply website/docs/r/monitor_data_collection_rule_association.html.markdown:129]: I think maybe this description is better
```suggestion
-> **NOTE** If `data_collection_endpoint_id` is specified, this data collection rule association must be named 'configurationAccessEndpoint'. 
```

File: website/docs/r/monitor_data_collection_rule_association.html.markdown (hunk 2 above)
Comment (katbyte) [reply website/docs/r/monitor_data_collection_rule_association.html.markdown:117]: they it sounds like this property should be optional and only required when data_collection_rule_id is configured, and if it is we populate it with configurationAccessEndpoint ?

**Review by katbyte — APPROVED** [review]
LGTM 🦺

File: website/docs/r/monitor_data_collection_rule_association.html.markdown [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+# associate to a Data Collection Rule
+resource "azurerm_data_collection_rule_association" "example1" {
+  name                    = "example1-dcra"
+  target_resource_id      = azurerm_linux_virtual_machine.example.id
+  data_collection_rule_id = azurerm_monitor_data_collection_rule.example.id
+  description             = "example"
+}
+
+# associate to a Data Collection Endpoint
+resource "azurerm_data_collection_rule_association" "example1" {
```
Comment (cschar) [inline website/docs/r/monitor_data_collection_rule_association.html.markdown:104]: should this be named example2? I tried pasting the code from https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/monitor_data_collection_rule_association and hit a 'example1' already exists error from the resource above defined on line 96

Comment (cschar) [reply website/docs/r/monitor_data_collection_rule_association.html.markdown:104]: also line 106, `data_collection_endpoint_id = azurerm_monitor_data_collection_endpoint.test.id` ?

Comment (teowa) [reply website/docs/r/monitor_data_collection_rule_association.html.markdown:104]: Thanks for pointing this out, I have submitted another PR #18536 to fix this.

**Pull request merged**